File: epi_judge_python_solutions/insert_operators_in_string.py
import functools

from test_framework import generic_test


def expression_synthesis(digits, target):
    def directed_expression_synthesis(digits, current_term):
        def evaluate():
            intermediate_operands = []
            operand_it = iter(operands)
            intermediate_operands.append(next(operand_it))
            # Evaluates '*' first.
            for oper in operators:
                if oper == '*':
                    product = intermediate_operands[-1] * next(operand_it)
                    intermediate_operands[-1] = product
                else:  # oper == '+'.
                    intermediate_operands.append(next(operand_it))
            # Evaluates '+' second.
            return sum(intermediate_operands)

        current_term = current_term * 10 + digits[0]
        if len(digits) == 1:
            operands.append(current_term)
            if evaluate() == target:  # Found a match.
                return True
            del operands[-1]
            return False

        # No operator.
        if directed_expression_synthesis(digits[1:], current_term):
            return True
        # Tries multiplication operator '*'.
        operands.append(current_term)
        operators.append('*')
        if directed_expression_synthesis(digits[1:], 0):
            return True
        del operands[-1]
        del operators[-1]
        # Tries addition operator '+'.
        operands.append(current_term)
        # First check feasibility of plus operator.
        if target - evaluate() <= functools.reduce(lambda val, d: val * 10 + d,
                                                   digits[1:], 0):
            operators.append('+')
            if directed_expression_synthesis(digits[1:], 0):
                return True
            del operators[-1]
        del operands[-1]
        return False

    operands, operators = [], []
    return directed_expression_synthesis(digits, 0)


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main("insert_operators_in_string.py",
                                       "insert_operators_in_string.tsv",
                                       expression_synthesis))

File: epi_judge_cpp_solutions/insert_operators_in_string.cc
#include <memory>
#include <numeric>
#include <stack>
#include <vector>

#include "test_framework/generic_test.h"

using std::accumulate;
using std::make_unique;
using std::stack;
using std::vector;

bool DirectedExpressionSynthesis(const vector<int>&, int, int, int,
                                 vector<int>*, vector<char>*);
int Evaluate(const vector<int>&, const vector<char>&);

bool ExpressionSynthesis(const vector<int>& digits, int target) {
  return DirectedExpressionSynthesis(digits, target, 0, 0,
                                     make_unique<vector<int>>().get(),
                                     make_unique<vector<char>>().get());
}

bool DirectedExpressionSynthesis(const vector<int>& digits, int target,
                                 int current_term, int offset,
                                 vector<int>* operands,
                                 vector<char>* operators) {
  current_term = current_term * 10 + digits[offset];
  if (offset == size(digits) - 1) {
    operands->emplace_back(current_term);
    if (Evaluate(*operands, *operators) == target) {  // Found a match.
      return true;
    }
    operands->pop_back();
    return false;
  }

  // No operator.
  if (DirectedExpressionSynthesis(digits, target, current_term, offset + 1,
                                  operands, operators)) {
    return true;
  }
  // Tries multiplication operator '*'.
  operands->emplace_back(current_term), operators->emplace_back('*');
  if (DirectedExpressionSynthesis(digits, target, /*current_term=*/0,
                                  offset + 1, operands, operators)) {
    return true;
  }
  operands->pop_back(), operators->pop_back();
  // Tries addition operator '+'.
  operands->emplace_back(current_term);
  // First check feasibility of plus operator.
  if (target - Evaluate(*operands, *operators) <=
      accumulate(begin(digits) + offset + 1, end(digits), 0,
                 [](int val, int d) { return val * 10 + d; })) {
    operators->emplace_back('+');
    if (DirectedExpressionSynthesis(digits, target, /*current_term=*/0,
                                    offset + 1, operands, operators)) {
      return true;
    }
    operators->pop_back();
  }
  operands->pop_back();
  return false;
}

int Evaluate(const vector<int>& operands, const vector<char>& operators) {
  stack<int> intermediate_operands;
  int operand_idx = 0;
  intermediate_operands.push(operands[operand_idx++]);
  // Evaluates '*' first.
  for (char oper : operators) {
    if (oper == '*') {
      int product = intermediate_operands.top() * operands[operand_idx++];
      intermediate_operands.pop();
      intermediate_operands.push(product);
    } else {  // oper == '+'.
      intermediate_operands.push(operands[operand_idx++]);
    }
  }

  // Evaluates '+' second.
  int sum = 0;
  while (!empty(intermediate_operands)) {
    sum += intermediate_operands.top();
    intermediate_operands.pop();
  }
  return sum;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"digits", "target"};
  return GenericTestMain(args, "insert_operators_in_string.cc",
                         "insert_operators_in_string.tsv", &ExpressionSynthesis,
                         DefaultComparator{}, param_names);
}
